### ISSUE TITLE
CheriAddressingModeFolder cannot fold constant addresses

### ISSUE THREAD

**Issue by Jerryxia32** [CTSRD-CHERI/llvm#251]
Any attempt to load/store via a capability at a constant address that fits in the immediate field will segfault. A trivial example is:
```
void
func() {
    void*__capability testCap = *((void*__capability volatile*)0x1000UL);
}
```
The reason is that CheriAddressingModeFolder cannot fold the following two instructions. Passing -mllvm -disable-cheri-addressing-mode-folder can avoid the segfault.
```
daddiu  $1, $zero, 4096
clc $c1, $1, 0($c0)
```
After some digging, if I understand this correctly, I think the bug is around line 165, CheriAddressingModeFolder.cpp:
```
if (tryToFoldAdd(Op, I.getOperand(1).getReg(), RI, AddInst, offset)) {
  Adds.insert(AddInst);
  I.getOperand(1).setImm(offset);
  I.getOperand(1).setReg(Mips::ZERO_64);
}
```
The first getOperand(1) should be (2) I think. It surprises me because it has not been found till now. I guess nobody does capability loads and stores at constant addresses unless this is really intended, and the chance of the address being able to fit in the immediate field is even rarer.

**Comment by RichardsonAlex** [comment]
This crash appears to be fixed in the dev branch however, it is still not folding the instructions.

**Comment by RichardsonAlex** [comment]
It appears the reason for it not being folded is that LLVM generates a `csetoffset` which is not handled by the addressing mode folder:

```
        cfromptr	$c2, $c0, $zero
	daddiu	$1, $zero, 4096
	csetoffset	$c2, $c2, $1
	clc	$c2, $zero, 0($c2)
```

**Comment by davidchisnall** [comment]
@Jerryxia32 Your fix does look correct, please commit it.
@RichardsonAlex We can't fold `csetoffset` in general, because we don't have an equivalent addressing mode.  In this particular case, we definitely can't because `cfromptr` will almost always give a capability with a non-zero offset.

**Comment by Jerryxia32** [comment]
Fixed in dev branch, commit b6680e9e29feb7e0fe92fc607745027ae033797f.

**Comment by Jerryxia32** [comment]
I should have mentioned that the crash happened with hybrid ABI.
